### ISSUE TITLE
when an app wants to add document metadata during its processing

### ISSUE THREAD

**Issue by keighrim** [clamsproject/mmif#134]
For instance, video processing apps that produce `TimeFrame` or `TimePoint` annotations can use frame number as the *unit* for those annotations. We talked about this in yesterday's meeting, and I thought if we encode the frame rate of the primary video file along with the frame numbers it would be helpful for downstream apps or the consumer of the MMIF file to convert the frame numbers into timecodes, by saving a tidious step of opening up the file (e.g. by `ffmepg`) and reading the fps information from the header. One question raised was where to put that information. Intuitively I thought that should be thrown into metadata of the `VideoDocument` objects, but I later realized that a `Document` object is supposed to be read-only. So next candidate would be `contains` metadata in the view. That would work, and in a sense, it seems to provide even better portability, but at the same time, it seems a bit off. I wonder how others think on this issue, and in more general, adding other possible instances of non-intrusive addition of metadata to the primary documents.

**Comment by marcverhagen** [comment]
Hmmmm. This is a question on whether something is part of the properties of a document or part of an annotation. 

- Document property. In this case it does make sense to have it stored on the VideDocument, but we would indeed have to add the frame rate and other pieces of information when the document object is created. I do think it would be good to store all those things that are required or useful for later processing.

- Annotation. Not sure how to encode this in the `contains` section. Under what annotation type will it go? It would be weird to say that a view contains instances of `VideoDocument` and say stuff about metadata and then not have any video documents in the view. Also, when we say metadata we mean properties that hold for all annotations of a particular type, which is different from the more normal sense of metadata (maybe we should have used a different name). 

And here is a slight of hand we could use. I think conceptually we said that views and annotations in the view should not be changed, and this was (I think) implemented by making views and annotations read-only. But we never really said anything about the properties of a source being read-only. Probably hell on earth to implement this though and I am probably not the only one who would dislike it when documents to have different statuses depending on where they occur.

**Comment by keighrim** [comment]
Another use case of such insertion of metadata that came up in today's meeting, brought by Kelley, is the screen canvas size (total height/width in pixel) of `VideoDocument`s that can be quickly utilized by per-frame image analysis apps.

**Comment by keighrim** [comment]
So Marc and I talked about this earlier this week, and we concluded that the best practice would be adding these kinds of information as `annotation` objects using [`Annotation`](https://mmif.clams.ai/0.4.0/vocabulary/Annotation/) `@type`, while keeping source data in `Document` objects untouched. Currently, the `Annotation` vocab type is there as an abstract, umbrella super class for other annotation types, but we can re-purpose it for concrete usage. 

For example, given a source MMIF 
``` json
{
  "metadata": {
    "mmif": "http://mmif.clams.ai/0.4.0"
  },
  "documents": [
    {
      "@type": "http://mmif.clams.ai/0.4.0/vocabulary/VideoDocument",
      "properties": {
        "mime": "video",
        "id": "d1",
        "location": "file:///a_video.mp4"
      }
    }
  ],
  "views": []
}
```

A video-processing app can add a simple `Annotation` of framerate along with its processing results. 
```json
{
  "metadata": {
    "mmif": "http://mmif.clams.ai/0.4.0"
  },
  "documents": [
    {
      "@type": "http://mmif.clams.ai/0.4.0/vocabulary/VideoDocument",
      "properties": {
        "mime": "video",
        "id": "d1",
        "location": "file:///a.mp4"
      }
    }
  ],
  "views": [
    {
      "id": "v_0",
      "metadata": {
        "timestamp": "some-time",
        "app": "some-video-app",
        "contains": {
          "http://mmif.clams.ai/0.4.0/vocabulary/Annotation": { # will hold general K-V pairs for document-level information/metadata
            "document": "d1"
          }
          "http://mmif.clams.ai/0.4.0/vocabulary/TimeFrame": {
            "timeUnit": "frames",
            "document": "d1"
          }
        }
      },
      "annotations": [
        {
          "@type": "http://mmif.clams.ai/0.4.0/vocabulary/Annotation",
          "properties": {
            "id": "ann_1",
            "framesPerSecond": 29.97 # we probably want to have a set of these key names pre-compiled for common metadata (fps, canvas size, ...) 
          }
        },
        {
          "@type": "http://mmif.clams.ai/0.4.0/vocabulary/TimeFrame",
          "properties": {
            "start": 100,
            "end": 1000,
            "frameType": "bars-and-tones",
            "id": "tf_1"
          }
        },
        # and more ...
      ]
    }
  ]
}

I am pretty sure that having a "capital" `Annotation` annotation type will bring quite a bit of communicative confusion in the future but other than that, I don't see a significant flaw in this approach, at least from a technical point. Let us know what others think.

**Comment by keighrim** [comment]
Closed via #191.
